### PR TITLE
Have .dockerignore support Dockerfile/.dockerignore

### PR DESCRIPTION
**Pull request by ybogdanov**
### Problem
Sometimes you have to add `Dockerfile` itself to `.dockerignore`, in cases like this https://github.com/docker/docker/issues/8330

For that purpose, the following changes were added to Docker: https://github.com/docker/docker/commit/6d801a3caa54ad7ef574bc426aa1ffc412c5af82

It became critical for me as well, so I decided to implement it also for `go-dockerclient`.

### Solution
Docker client will read `.dockerignore` and whilelist `Dockerfile` or `.dockerignore` to make sure they will be transferred to a server. Server is reading `.dockerignore` again, and removing those files after capturing them in memory https://github.com/docker/docker/blob/596e91638cbe8b583dff4fdfc5a52e9d3840b20f/builder/evaluator.go#L240

### CODE REVIEW

File: tar.go [hunk 1]
```diff
@@ -17,17 +17,36 @@ import (
...
+	// If .dockerignore mentions .dockerignore or the Dockerfile
+	// then make sure we send both files over to the daemon
+	// because Dockerfile is, obviously, needed no matter what, and
+	// .dockerignore is needed to know if either one needs to be
+	// removed.  The deamon will remove them for us, if needed, after it
+	// parses the Dockerfile.
+	//
+	// https://github.com/docker/docker/issues/8330
+	//
+	if keepThem, _ := fileutils.Matches(".dockerignore", excludes); keepThem {
```
Comment (peter-edge) [inline tar.go]: check error, do not ignore, this will fail errcheck

File: tar.go [hunk 2]
```diff
@@ -17,17 +17,36 @@ import (
...
+		includes = append(includes, ".dockerignore")
+	}
+	if keepThem, _ := fileutils.Matches(dockerfileName, excludes); keepThem && dockerfileName != "" {
```
Comment (peter-edge) [inline tar.go]: same

File: tar.go [hunk 3]
```diff
@@ -17,17 +17,36 @@ import (
 	"github.com/docker/docker/pkg/fileutils"
 )
 
-func createTarStream(srcPath string) (io.ReadCloser, error) {
+func createTarStream(srcPath, dockerfileName string) (io.ReadCloser, error) {
```
Comment (peter-edge) [inline tar.go]: nit: probably would change to dockerfilePath

**Comment by fsouza** [comment]
Hi @ybogdanov, thanks for contributing!

@peter-edge thanks for reviewing it, it's good to merge! :-)

**Pull request merged**